File: wisp/wisp-cli/Sources/wisp/Logger.swift
import Foundation
import WispAI

// MARK: - Session Logger

/// Writes a JSONL session log to ~/.pi/agent/sessions/<timestamp>_<uuid>.jsonl
/// Format mirrors the original pi session-manager output (version 1 for wisp-cli).
actor SessionLogger {

    private let fileHandle: FileHandle
    private var lastMessageId: String?

    private init(fileHandle: FileHandle) {
        self.fileHandle = fileHandle
    }

    // MARK: - Factory

    static func open(model: Model) throws -> SessionLogger {
        let sessionsDir = SettingsManager.agentDir + "/sessions"
        try FileManager.default.createDirectory(
            atPath: sessionsDir,
            withIntermediateDirectories: true,
            attributes: nil
        )

        let sessionId  = UUID().uuidString.lowercased()
        let timestamp  = ISO8601DateFormatter().string(from: Date())
        let safestamp  = timestamp.replacing(":", with: "-").replacing(".", with: "-")
        let filename   = "\(safestamp)_\(sessionId).jsonl"
        let path       = sessionsDir + "/" + filename

        FileManager.default.createFile(atPath: path, contents: nil)
        guard let fh = FileHandle(forWritingAtPath: path) else {
            throw LoggerError.cannotOpenFile(path)
        }

        let logger = SessionLogger(fileHandle: fh)

        let header: [String: Any] = [
            "type":      "session",
            "version":   1,
            "id":        sessionId,
            "timestamp": timestamp,
            "cwd":       FileManager.default.currentDirectoryPath,
            "tool":      "wisp",
            "model":     model.id,
            "provider":  model.provider,
        ]
        try logger.appendEntry(header)
        return logger
    }

    // MARK: - Public API

    func logUserMessage(text: String) throws {
        let id = UUID().uuidString.lowercased()
        let entry: [String: Any] = [
            "type":      "message",
            "id":        id,
            "parentId":  NSNull(),
            "timestamp": ISO8601DateFormatter().string(from: Date()),
            "message": [
                "role":    "user",
                "content": text,
            ] as [String: Any],
        ]
        try appendEntry(entry)
        lastMessageId = id
    }

    func logAssistantMessage(_ msg: AssistantMessage) throws {
        let id = UUID().uuidString.lowercased()
        var contentArray: [[String: Any]] = []
        for part in msg.content {
            switch part {
            case .text(let t):
                contentArray.append(["type": "text", "text": t.text])
            case .thinking(let t):
                contentArray.append(["type": "thinking", "thinking": t.thinking])
            case .toolCall(let tc):
                contentArray.append([
                    "type":      "tool_call",
                    "id":        tc.id,
                    "name":      tc.name,
                    "arguments": tc.arguments,
                ])
            }
        }

        var msgDict: [String: Any] = [
            "role":       "assistant",
            "model":      msg.model,
            "provider":   msg.provider,
            "stopReason": msg.stopReason.rawValue,
            "usage": [
                "input":  msg.usage.input,
                "output": msg.usage.output,
                "cost":   msg.usage.cost.total,
            ] as [String: Any],
            "content": contentArray,
        ]
        if let err = msg.errorMessage { msgDict["errorMessage"] = err }
        if !msg.diagnostics.isEmpty {
            msgDict["diagnostics"] = msg.diagnostics.map { d -> [String: Any] in
                var de: [String: Any] = [
                    "type":      d.type,
                    "timestamp": ISO8601DateFormatter().string(from: d.timestamp),
                ]
                if let e = d.error {
                    var errDict: [String: Any] = ["message": e.message]
                    if let n = e.name { errDict["name"] = n }
                    if let c = e.code { errDict["code"] = c }
                    de["error"] = errDict
                }
                if !d.details.isEmpty { de["details"] = d.details }
                return de
            }
        }

        let parentId: Any = lastMessageId as Any? ?? NSNull()
        let entry: [String: Any] = [
            "type":      "message",
            "id":        id,
            "parentId":  parentId,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
            "message":   msgDict,
        ]
        try appendEntry(entry)
        lastMessageId = id
    }

    func close() {
        try? fileHandle.close()
    }

    // MARK: - Private

    private func appendEntry(_ dict: [String: Any]) throws {
        guard let data = try? JSONSerialization.data(
            withJSONObject: dict,
            options: [.sortedKeys]
        ) else { throw LoggerError.serializationFailed }

        var line = data
        line.append(0x0A) // newline
        try fileHandle.write(contentsOf: line)
    }
}

enum LoggerError: Error {
    case cannotOpenFile(String)
    case serializationFailed
}

// String helper — available on Linux/macOS 13+.
private extension String {
    func replacing(_ target: Character, with replacement: Character) -> String {
        map { $0 == target ? replacement : $0 }.reduce("") { $0 + String($1) }
    }
}

File: wisp/wisp/logger.go
package main

import (
	"crypto/rand"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"wispai"
)

// SessionLogger appends JSONL entries to a session log file under
// ~/.pi/agent/sessions/.
type SessionLogger struct {
	file          *os.File
	lastMessageID string
}

// NewSessionLogger creates the sessions directory if needed, opens a new
// JSONL file, and writes the session header entry.
func NewSessionLogger(model wispai.Model) (*SessionLogger, error) {
	sessionsDir := wispai.AgentDir() + "/sessions"
	if err := os.MkdirAll(sessionsDir, 0o755); err != nil {
		return nil, err
	}

	sessionID := newUUID()
	now := time.Now().UTC()
	ts := now.Format("2006-01-02T15:04:05.000Z")
	safe := strings.NewReplacer(":", "-", ".", "-").Replace(ts)
	path := fmt.Sprintf("%s/%s_%s.jsonl", sessionsDir, safe, sessionID)

	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return nil, err
	}

	cwd, _ := os.Getwd()
	lg := &SessionLogger{file: f}
	if err := lg.appendEntry(map[string]any{
		"type":      "session",
		"version":   1,
		"id":        sessionID,
		"timestamp": now.Format(time.RFC3339Nano),
		"cwd":       cwd,
		"tool":      "wisp",
		"model":     model.ID,
		"provider":  model.Provider,
	}); err != nil {
		f.Close()
		return nil, err
	}
	return lg, nil
}

// LogUserMessage appends a user message entry.
func (lg *SessionLogger) LogUserMessage(text string) error {
	id := newUUID()
	err := lg.appendEntry(map[string]any{
		"type":      "message",
		"id":        id,
		"parentId":  nil,
		"timestamp": time.Now().UTC().Format(time.RFC3339Nano),
		"message": map[string]any{
			"role":    "user",
			"content": text,
		},
	})
	if err == nil {
		lg.lastMessageID = id
	}
	return err
}

// LogAssistantMessage appends an assistant message entry.
func (lg *SessionLogger) LogAssistantMessage(msg *wispai.AssistantMessage) error {
	content := make([]map[string]any, 0, len(msg.Content))
	for _, p := range msg.Content {
		switch p.Type {
		case "text":
			content = append(content, map[string]any{"type": "text", "text": p.Text})
		case "thinking":
			content = append(content, map[string]any{"type": "thinking", "thinking": p.Thinking})
		case "toolCall":
			content = append(content, map[string]any{
				"type": "tool_call", "id": p.ID, "name": p.Name, "arguments": p.Arguments,
			})
		}
	}

	msgDict := map[string]any{
		"role":       "assistant",
		"model":      msg.Model,
		"provider":   msg.Provider,
		"stopReason": string(msg.StopReason),
		"usage": map[string]any{
			"input":  msg.Usage.Input,
			"output": msg.Usage.Output,
			"cost":   msg.Usage.Cost.Total,
		},
		"content": content,
	}
	if msg.ErrorMessage != "" {
		msgDict["errorMessage"] = msg.ErrorMessage
	}
	if len(msg.Diagnostics) > 0 {
		diags := make([]map[string]any, 0, len(msg.Diagnostics))
		for _, d := range msg.Diagnostics {
			de := map[string]any{
				"type":      d.Type,
				"timestamp": d.Timestamp.UTC().Format(time.RFC3339Nano),
			}
			if d.Error != nil {
				e := map[string]any{"message": d.Error.Message}
				if d.Error.Name != "" {
					e["name"] = d.Error.Name
				}
				if d.Error.Code != "" {
					e["code"] = d.Error.Code
				}
				de["error"] = e
			}
			if len(d.Details) > 0 {
				de["details"] = d.Details
			}
			diags = append(diags, de)
		}
		msgDict["diagnostics"] = diags
	}

	var parentID any
	if lg.lastMessageID != "" {
		parentID = lg.lastMessageID
	}

	id := newUUID()
	err := lg.appendEntry(map[string]any{
		"type":      "message",
		"id":        id,
		"parentId":  parentID,
		"timestamp": time.Now().UTC().Format(time.RFC3339Nano),
		"message":   msgDict,
	})
	if err == nil {
		lg.lastMessageID = id
	}
	return err
}

// Close flushes and closes the underlying file.
func (lg *SessionLogger) Close() error { return lg.file.Close() }

func (lg *SessionLogger) appendEntry(d map[string]any) error {
	b, err := json.Marshal(d)
	if err != nil {
		return err
	}
	b = append(b, '\n')
	_, err = lg.file.Write(b)
	return err
}

func newUUID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant RFC 4122
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
